services/rsources/stats_collector.go

```diff
@@ -0,0 +1,152 @@
+	BeginProcessing(jobs []*jobsdb.JobT)
+
+	// JobStatusesUpdated captures outgoing job statistics.
+	// A call to BeginProcessing must preceed a call to this method,
+	// so that all necessary indices can been created, since a JobStatus
+	// doesn't carry all necessary job metadata such as jobRunId, taskRunId, etc.
+	JobStatusesUpdated(jobStatuses []*jobsdb.JobStatusT)
+
+	// Publish publishes statistics
+	Publish(ctx context.Context, tx *sql.Tx) error
+}
+
+// NewStatsCollector creates a new stats collector
+func NewStatsCollector(service JobService) StatsCollector {
+	return &statsCollector{
+		service:            service,
+		jobIdsToStatsIndex: map[int64]*Stats{},
+		statsIndex:         map[statKey]*Stats{},
+	}
+}
+
+type statKey struct {
+	jobRunId string
+	JobTargetKey
+}
+
+var _ StatsCollector = (*statsCollector)(nil)
+
+type statsCollector struct {
+	processing         bool
+	service            JobService
+	jobIdsToStatsIndex map[int64]*Stats
+	statsIndex         map[statKey]*Stats
+}
+
+func (r *statsCollector) JobsStored(jobs []*jobsdb.JobT) {
+	r.buildStats(jobs, nil, true)
+}
+
+func (r *statsCollector) JobsStoredWithErrors(jobs []*jobsdb.JobT, failedJobs map[uuid.UUID]string) {
+	r.buildStats(jobs, failedJobs, true)
+}
+
+func (r *statsCollector) BeginProcessing(jobs []*jobsdb.JobT) {
+	r.buildStats(jobs, nil, false)
+	r.processing = true
+}
+
+func (r *statsCollector) JobStatusesUpdated(jobStatuses []*jobsdb.JobStatusT) {
+	if !r.processing {
+		panic(fmt.Errorf("cannot update job statuses without having previously called BeginProcessing"))
+	}
+	if len(r.jobIdsToStatsIndex) == 0 {
+		return
+	}
+	for i := range jobStatuses {
+		jobStatus := jobStatuses[i]
+		stats, ok := r.jobIdsToStatsIndex[jobStatus.JobID]
+		if ok {
+			switch jobStatus.JobState {
+			case jobsdb.Succeeded.State:
+				stats.Out++
+			case jobsdb.Aborted.State:
+				stats.Failed++
+			}
+		}
+	}
+}
+
+func (r *statsCollector) Publish(ctx context.Context, tx *sql.Tx) error {
+	if r.service == nil {
+		return fmt.Errorf("No JobService provided during initialization")
+	}
+	for k, v := range r.statsIndex {
+		if v.Failed+v.In+v.Out == 0 {
+			continue
+		}
+		err := r.service.IncrementStats(ctx, tx, k.jobRunId, k.JobTargetKey, *v)
+		if err != nil {
+			return err
+		}
+	}
+
+	// reset state so that the collector can be
+	// reused for another stats collecting cycle
+	r.processing = false
+	r.jobIdsToStatsIndex = map[int64]*Stats{}
+	r.statsIndex = map[statKey]*Stats{}
+
+	return nil
+}
+
+const (
+	source_job_run_id  = "source_job_run_id"
+	source_task_run_id = "source_task_run_id"
+	source_id          = "source_id"
+	destination_id     = "destination_id"
+)
+
+func (r *statsCollector) buildStats(jobs []*jobsdb.JobT, failedJobs map[uuid.UUID]string, incrementIn bool) { // skipcq: RVV-A0005
+	for i := range jobs {
+		job := jobs[i]
+		if _, ok := failedJobs[job.UUID]; ok {
+			continue
+		}
+		jobRunId := gjson.GetBytes(job.Parameters, source_job_run_id).Str
```

Comment:
Would it? In which scenario, when the job has a source_job_run_id parameter or not?
The assumption here is that it is less likely for a job to not be from sources, thus only paying the small overhead of `gjson.GetBytes`. For the sources scenario, I don't really know if it is more performant to use `gjson.GetBytes` 3 times vs unmarshalling to json, I assume though that it will be.

Reply:
From my memory gjson is less performant and parameters have a defined structure so assumed unMarshal would be more performant. But yeah a benchmark would help

Reply:
Here are some results. Even though results will vary depending on the actual json, the difference in performance between the two methods doesn't appear to be considerable, wdyt?
```
BenchmarkParamsParsing/parse_params_using_gjson.GetBytes_3_times-10         	 2830051	       409.1 ns/op	      56 B/op	       3 allocs/op
BenchmarkParamsParsing/parse_params_using_jsoniter.Unmarshall-10            	 2552539	       472.6 ns/op	     104 B/op	       4 allocs/op
```

Reply:
Yeah true.. there isn't major difference between the two. Resolving this comment